receiver/mongodbreceiver/scraper.go

```diff
@@ -0,0 +1,171 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mongodbreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbreceiver"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-version"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.opentelemetry.io/collector/receiver/scrapererror"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbreceiver/internal/metadata"
+)
+
+const instrumentationLibraryName = "otelcol/mongodb"
+
+type mongodbScraper struct {
+	logger       *zap.Logger
+	config       *Config
+	client       client
+	mongoVersion *version.Version
+	mb           *metadata.MetricsBuilder
+}
+
+func newMongodbScraper(logger *zap.Logger, config *Config) *mongodbScraper {
+	return &mongodbScraper{
+		logger: logger,
+		config: config,
+		mb:     metadata.NewMetricsBuilder(config.Metrics),
+	}
+}
+
+func (s *mongodbScraper) start(ctx context.Context, _ component.Host) error {
+	clientLogger := s.logger.Named("mongo-scraper")
+	c := NewClient(s.config, clientLogger)
+	s.client = c
+
+	if err := s.client.Connect(ctx); err != nil {
+		s.logger.Error("unable to connect to mongo instance", zap.Error(err))
+	}
+
+	vr, err := s.client.GetVersion(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to get a version from the mongo instance: %w", err)
+	}
+	v, err := version.NewVersion(*vr)
+	if err != nil {
+		return fmt.Errorf("unable to parse version as semantic version %s: %w", *vr, err)
+	}
+
+	s.mongoVersion = v
```

Comment:
`start` should not fail based on inability to reach the target. If we can't get the version here, we should still try again later when `scrape` is called.

I dug into the docs a bit and it looks like the mongo driver's `Connect` function only fails if there is something invalid in the configuration of the client. It does not actually attempt to connect to the target. So actually we _should_ return an error if `Connect` does.

The problem, is that we've applied a wrapper around the real client's `Connect`, which actually does require a connection to be established. This depends on the `Ping` function, which presents a few issues here. Notably, Ping will hang until the connection times out. Since the default timeout is 1 minute, the entire collector's startup routine can be blocked on this for an unreasonable amount of time. A minute might be a reasonable timeout for the `scrape` function, but not `start`. 

The other thing I noticed about the client is that it is a "smart" client. That is, when you attempt to make any request, it will figure out if it needs to establish a connection or not. 

So here's what I think we need to do:
1. Remove the wrapper around `client.Connect`. We don't actually want to add anything to it, just call it directly.
2. If `client.Connect` returns an error, `start` should as well. (Again, this does not mean that the target was unavailable - it means that the configuration is invalid)
3. You can still attempt to get the version, but do not return an error if it fails.
4. In `start`, check if the version was previously found. If not, try to get it. If that fails, return an error from scrape, because we can't safely collect without knowing it, and the target appears to be inaccessible at the moment anyways.

Reply:
Roughly:
```go
func start() {

func (s *mongodbScraper) start(ctx context.Context, _ component.Host) error {
	clientLogger := s.logger.Named("mongo-scraper")
	c := NewClient(s.config, clientLogger)
	s.client = c

	/* Importantly, this should be calling Connect directly, not through a wrapper */
	if err := s.client.Connect(ctx); err != nil {
		return fmt.Errorf(..., err)
	}

	/* I think it's ok to have the client also parse the version. */
	v, err := s.client.GetVersion(ctx)
	if err != nil {
		s.logger.Errorf("failed to get version at start: ...")
	}

	return nil
}

func (s *mongodbScraper) scrape(ctx context.Context) (pdata.Metrics, error) {
	if s.client == nil {
		return pdata.NewMetrics(), errors.New("no client was initialized before calling scrape")
	}

	if s.mongoVersion == nil {
		v, err := s.client.GetVersion(ctx)
		if err != nil {
			return fmt.Errorf(...)
		}
	// ok to proceed
}

```

Reply:
One question I have about the suggestion to use the raw `Connect` call is that it initially did one thing important thing in that it sets the underlying client for other calls. If we did the raw call you’d end up getting panics with any later attempt at GetVersion or any other call like. Would you be fine if we have a connect wrapper just no `Ping` call to validate the connection?

Reply:
Yep, that's fine. Just need to error only when the underlying one does.